pkg/jobrunaggregator/jobruntestcaseanalyzer/analyzer.go

```diff
@@ -0,0 +1,471 @@
+	for i := range jobs {
+		job := jobs[i]
+		var jobRunLocator jobrunaggregatorlib.JobRunLocator
+
+		if len(o.payloadTag) > 0 {
+			jobRunLocator = jobrunaggregatorlib.NewPayloadAnalysisJobLocatorForReleaseController(
+				job.JobName,
+				o.payloadTag,
+				o.jobRunStartEstimate,
+				o.ciDataClient,
+				o.ciGCSClient,
+				o.gcsClient,
+				"origin-ci-test",
+			)
+		}
+		if len(o.payloadInvocationID) > 0 {
+			jobRunLocator = jobrunaggregatorlib.NewPayloadAnalysisJobLocatorForPR(
+				job.JobName,
+				o.payloadInvocationID,
+				jobrunaggregatorlib.PayloadInvocationIDLabel,
+				o.jobRunStartEstimate,
+				o.ciDataClient,
+				o.ciGCSClient,
+				o.gcsClient,
+				"origin-ci-test",
+				(*o.jobGCSPrefixes)[i].gcsPrefix,
+			)
+		}
+
+		fmt.Printf("  launching findJobRunsWithRetry for %q\n", job.JobName)
+
+		waitGroup.Add(1)
+
+		go func() {
+			defer waitGroup.Done()
+			jobRuns, err := o.findJobRunsWithRetry(ctx, job.JobName, jobRunLocator)
+			if err == nil {
+				resultCh <- jobRuns
+			}
+		}()
+	}
+	waitGroup.Wait()
+	close(resultCh)
+
+	// drain the result channel first
+	for jobRuns := range resultCh {
+		jobRunsToReturn = append(jobRunsToReturn, jobRuns...)
+	}
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+		break
+	}
+	return jobRunsToReturn, nil
+}
+
+func (o *JobRunTestCaseAnalyzerOptions) runTestCaseCheckers(ctx context.Context,
+	finishedJobRuns []jobrunaggregatorapi.JobRunInfo, unfinishedJobRuns []jobrunaggregatorapi.JobRunInfo) *junit.TestSuite {
+	suiteName := "payload-cross-jobs"
+	topSuite := &junit.TestSuite{
+		Name:      suiteName,
+		TestCases: []*junit.TestCase{},
+	}
+
+	allJobRuns := append(finishedJobRuns, unfinishedJobRuns...)
+	jobRunJunitMap := map[jobrunaggregatorapi.JobRunInfo]*junit.TestSuites{}
+	for i := range allJobRuns {
+		jobRun := allJobRuns[i]
+
+		testSuites, err := jobRun.GetCombinedJUnitTestSuites(ctx)
+		if err != nil {
+			continue
+		}
+		jobRunJunitMap[jobRun] = testSuites
+	}
+	for _, checker := range o.testCaseCheckers {
+		testSuite := checker.CheckTestCase(ctx, jobRunJunitMap)
+		topSuite.Children = append(topSuite.Children, testSuite)
+		topSuite.NumTests += testSuite.NumTests
+		topSuite.NumFailed += testSuite.NumFailed
+	}
+	return topSuite
+}
+
+func (o *JobRunTestCaseAnalyzerOptions) Run(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, o.timeout)
+	defer cancel()
+
+	matchID := o.payloadTag
+	if len(matchID) == 0 {
+		matchID = o.payloadInvocationID
+	}
+
+	outputDir := filepath.Join(o.workingDir, matchID)
+	if err := os.MkdirAll(outputDir, 0755); err != nil {
+		return fmt.Errorf("error creating output directory %q: %w", outputDir, err)
+	}
+
+	// if it hasn't been more than hour since the jobRuns started, the list isn't complete.
+	readyAt := o.jobRunStartEstimate.Add(1 * time.Hour)
+
+	// Our guess of the max time for jobs
+	durationToWait := o.timeout - 20*time.Minute
+	if durationToWait > (4*time.Hour + 15*time.Minute) {
```

Comment:
I would output a warning here if you're overriding the users requested timeout. This actually might be better off getting rejected in the CLI validation, as this seems like it might be doing something the user doesn't expect and won't notice.